File: plugins/plugin-slack/typescript/src/actions/listChannels.ts
import type {
  Action,
  ActionExample,
  ActionResult,
  Content,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import type { SlackService } from "../service";
import { SLACK_SERVICE_NAME } from "../types";

export const listChannels: Action = {
  name: "SLACK_LIST_CHANNELS",
  similes: [
    "LIST_SLACK_CHANNELS",
    "SHOW_CHANNELS",
    "GET_CHANNELS",
    "CHANNELS_LIST",
  ],
  description: "List available Slack channels in the workspace",
  validate: async (
    _runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
  ): Promise<boolean> => {
    return message.content.source === "slack";
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
  ): Promise<ActionResult | undefined> => {
    const slackService = await runtime.getService(SLACK_SERVICE_NAME) as SlackService;

    if (!slackService || !slackService.client) {
      await callback?.({
        text: "Slack service is not available.",
        source: "slack",
      });
      return { success: false, error: "Slack service not available" };
    }

    const channels = await slackService.listChannels({
      types: "public_channel,private_channel",
      limit: 100,
    });

    // Sort channels by name
    const sortedChannels = channels
      .filter((ch) => !ch.isArchived)
      .sort((a, b) => a.name.localeCompare(b.name));

    // Format channel list
    const channelList = sortedChannels.map((ch) => {
      const memberCount =
        ch.numMembers !== undefined ? ` (${ch.numMembers} members)` : "";
      const privateIndicator = ch.isPrivate ? " 🔒" : "";
      const topic = ch.topic?.value
        ? ` - ${ch.topic.value.slice(0, 50)}${ch.topic.value.length > 50 ? "..." : ""}`
        : "";
      return `• #${ch.name}${privateIndicator}${memberCount}${topic}`;
    });

    const response: Content = {
      text: `Found ${sortedChannels.length} channels:\n\n${channelList.join("\n")}`,
      source: message.content.source,
    };

    runtime.logger.debug(
      {
        src: "plugin:slack:action:list-channels",
        channelCount: sortedChannels.length,
      },
      "[SLACK_LIST_CHANNELS] Channels listed",
    );

    await callback?.(response);

    return {
      success: true,
      data: {
        channelCount: sortedChannels.length,
        channels: sortedChannels.map((ch) => ({
          id: ch.id,
          name: ch.name,
          isPrivate: ch.isPrivate,
          numMembers: ch.numMembers,
          topic: ch.topic?.value,
          purpose: ch.purpose?.value,
        })),
      },
    };
  },
  examples: [
    [
      {
        name: "{{user1}}",
        content: {
          text: "Show me all the channels in this workspace",
        },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I'll list all the available channels.",
          actions: ["SLACK_LIST_CHANNELS"],
        },
      },
    ],
    [
      {
        name: "{{user1}}",
        content: {
          text: "What channels can I join?",
        },
      },
      {
        name: "{{agent}}",
        content: {
          text: "Let me show you the available channels.",
          actions: ["SLACK_LIST_CHANNELS"],
        },
      },
    ],
  ] as ActionExample[][],
};

export default listChannels;

File: plugins/plugin-slack/rust/src/actions/list_channels.rs
//! List channels action for Slack.

use crate::service::SlackService;
use crate::types::SlackChannel;
use serde::{Deserialize, Serialize};

/// Action name
pub const LIST_CHANNELS_ACTION: &str = "SLACK_LIST_CHANNELS";

/// Action similes
pub const LIST_CHANNELS_SIMILES: &[&str] = &[
    "LIST_SLACK_CHANNELS",
    "SHOW_CHANNELS",
    "GET_CHANNELS",
];

/// Action description
pub const LIST_CHANNELS_DESCRIPTION: &str = "List available Slack channels in the workspace";

/// List channels result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListChannelsResult {
    pub success: bool,
    pub channels: Vec<ChannelInfo>,
    pub channel_count: usize,
    pub error: Option<String>,
}

/// Channel info for listing
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelInfo {
    pub id: String,
    pub name: String,
    pub is_private: bool,
    pub num_members: Option<i32>,
    pub topic: Option<String>,
    pub purpose: Option<String>,
}

impl From<SlackChannel> for ChannelInfo {
    fn from(ch: SlackChannel) -> Self {
        Self {
            id: ch.id,
            name: ch.name,
            is_private: ch.is_private,
            num_members: ch.num_members,
            topic: ch.topic.map(|t| t.value),
            purpose: ch.purpose.map(|p| p.value),
        }
    }
}

/// Execute the list channels action
pub async fn execute_list_channels(service: &SlackService) -> ListChannelsResult {
    match service.list_channels(Some("public_channel,private_channel"), Some(100)).await {
        Ok(channels) => {
            let channels: Vec<ChannelInfo> = channels
                .into_iter()
                .filter(|ch| !ch.is_archived)
                .map(ChannelInfo::from)
                .collect();
            
            let channel_count = channels.len();
            
            ListChannelsResult {
                success: true,
                channels,
                channel_count,
                error: None,
            }
        }
        Err(e) => ListChannelsResult {
            success: false,
            channels: vec![],
            channel_count: 0,
            error: Some(e.to_string()),
        },
    }
}
